### PR TITLE
fix: make sure SNI is changed with HTTP host

### LINKED ISSUE #3864: bug: Upstream SNI should also be rewritten when the upstream host is rewritten
### Issue description

Currently, when the upstream host is changed (by `proxy-rewrite plugin, or the Host rewrite policy in Upstream), the SNI passed to Upstream is not changed when the scheme is HTTPS. That might introduce some SSL handshaking problem, we may fix it.

### Environment

* apisix version (cmd: `apisix version`):
* OS (cmd: `uname -a`):
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
* apisix-dashboard version, if have:

### Minimal test code / Steps to reproduce the issue

1.
2.
3.

### PR DESCRIPTION
**Pull request by spacewander**
Move the variables under locations to avoid being reset by proxy-mirror.
Fix #3864
Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/cli/test_access_log.sh [hunk 1]
```diff
@@ -183,3 +183,44 @@ fi
...
+
+# APISIX
+if ! grep '"https://localhost" -' output.log; then
+    echo "failed: find upstream scheme"
+    cat output.log
+    exit 1
+fi
+
+# admin
+if ! grep '"http://localhost" localhost' output.log; then
```
Comment (tokers) [inline t/cli/test_access_log.sh:219]: Where does this log generate?

Comment (spacewander) [reply t/cli/test_access_log.sh:219]: `tail -n 2 logs/access.log > output.log`, a few lines above

File: t/cli/test_access_log.sh [hunk 2]
```diff
@@ -183,3 +183,44 @@ fi
...
+curl -k -i https://127.0.0.1:9180/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d \
+    '{"uri":"/apisix/admin/routes/1", "upstream":{"nodes":{"localhost:9180":1},"scheme":"https","type":"roundrobin","pass_host":"node"}}'
+
+curl -i http://127.0.0.1:9080/apisix/admin/routes/1
+sleep 4
+tail -n 2 logs/access.log > output.log
+
+# APISIX
+if ! grep '"https://localhost" -' output.log; then
+    echo "failed: find upstream scheme"
```
Comment (Firstsawyou) [inline t/cli/test_access_log.sh:213]: Prompt message, I think it would be better like this.
```
    echo "failed: failed to find the upstream scheme"
```

Comment (spacewander) [reply t/cli/test_access_log.sh:213]: The prompt is just the title of this test, see: https://github.com/apache/apisix/blob/a7e7d60e3b620410c42a0d773baaaf995b4926a3/t/cli/test_validate_config.sh#L31

Ignore the negative/positive word can make life easier.

Comment (spacewander) [reply t/cli/test_access_log.sh:213]: Anyway, I will tweak the message to make it clear.

**Review by tokers — APPROVED** [review]

**Review by moonming — APPROVED** [review]

**Pull request merged**